### ISSUE TITLE
Zenmap and Ndiff are python2 only

### ISSUE THREAD

**Issue by landgraf** [nmap/nmap#1176]
Python 2.X faces official EOL in 2020 and most of the Linux distributions started to drop python2 support already. To keep zenmap and ndiff they should be ported to python3.

**Comment by dmiller-nmap** [comment]
Duplicate of #665

**Comment by dmiller-nmap** [comment]
Sorry, thought we already had an issue for this, but I got the wrong thing (a PR instead).

**Comment by Vascom** [comment]
Any work in progress?

**Comment by dmiller-nmap** [comment]
@Vascom No, no work is in progress. We would accept a pull request if one was made that passed review. Some challenges that would have to be overcome:

1. Zenmap makes extensive use of `pygtk`, and I'm not sure that the equivalent Python 3 package is compatible. Nothing definite, just a concern that would have to be investigated.
2. We do packaging of Zenmap and Ndiff for Windows and OS X using `py2exe` and [`gtk-mac-bundler`](https://wiki.gnome.org/Projects/GTK+/OSX/Bundling), and those processes will require adjustment for Python 3 and whatever GTK+ library is compatible.

**Comment by samueloph** [comment]
I've received the following from a bug report on Debian:

> pygtk is unmaintained upstream. It has not had a release since GNOME 3
> was released in 2011.
> 
> The way forward is to port your app to use GObject Introspection
> bindings (and gtk3).
> 
> For more information on GObject Introspection see [1] and [2].
> 
> Please try to do this before the Buster release as we're going to
> try to remove pygtk this cycle.
> 
> If you have any question don't hesitate to ask.
> 
> [1] https://wiki.gnome.org/Projects/GObjectIntrospection
> [2] https://wiki.gnome.org/Projects/PyGObject

The bug report is here: https://bugs.debian.org/cgi-bin/bugreport.cgi?bug=885498

I believe that means you might want to switch to GObject Introspection while also dropping the python2 dependency.

Thanks

**Comment by a-andre** [comment]
See #1484 for a first step to support python3.

**Comment by BryanQuigley** [comment]
Can python2 support be dropped entirely?  Just jump straight to python3?

**Comment by landgraf** [comment]
> Can python2 support be dropped entirely? Just jump straight to python3?

It can be dropped altogether with nping and zenmap only. Porting gtk application from Python 2 to Python 3 is not simple task but patches are welcomed :)

**Comment by BryanQuigley** [comment]
I didn't even see that nping had python2 bits.    Does ndiff need to keep python2 support?

**Comment by landgraf** [comment]
> I didn't even see that nping had python2 bits. Does ndiff need to keep python2 support?

I meant ndiff and zenmap. Sorry

**Comment by BryanQuigley** [comment]
^^ #1807 Proposed port of ndiff - did everything but setup.py.

**Comment by XVilka** [comment]
Less than 2 months left for the old snake to be alive: https://pythonclock.org/

**Comment by BryanQuigley** [comment]
Zenmap has been removed from Debian unstable - and ndiff in Debian has been kept by merging my PR #1807 ahead of upstream.

I don't have time at the moment, but I'm not planning on trying Zenmap porting until ndiff is merged.

**Comment by BryanQuigley** [comment]
@landgraf @dmiller-nmap
Can either of you push for merge of #1807 ?  Thanks!

**Comment by fyodor** [comment]
I just wanted to note that there are some useful comments in the closed duplicate issue #2022 .  In particular, here is a workaround posted by @N0rbert for Ubuntu 20.04 LTS:

```
mkdir -p ~/Downloads/zenmap
cd ~/Downloads/zenmap

wget http://archive.ubuntu.com/ubuntu/pool/universe/p/pygtk/python-gtk2_2.24.0-6_amd64.deb
wget http://archive.ubuntu.com/ubuntu/pool/universe/n/nmap/zenmap_7.80+dfsg1-1build1_all.deb

sudo apt install ./*.deb
```

**Comment by fyodor** [comment]
I also wanted to note PR #1972  here.  That's a patch which tries to support Python3 while still retaining Python2 compatibility.

**Comment by fyodor** [comment]
I also wanted to note that there are useful comments in closed duplicate issue #1684   They noted that  the OpenSUSE Leap 15.1, Fedora 31, Ubuntu 20.04LTS and Kali 2020.2 also have Python3 as default and thus suffer for this problem.  As a workaround until we finish the Python3 port, @rrioux suggested editing /usr/bin/zenmap to directly use python2 by switching the top line to "#!/usr/bin/env python2".  Of course you need to install python2 and pygtk in that case too.

**Comment by necrose99** [comment]
http://zetcode.com/python/gtk/  lists a python3 port.

**Comment by ipatrol** [comment]
I suggest that this no longer be tagged "enhancement" and be considered a critical bug. Zenmap is considered unmaintained by distros because nothing has been done on this. If this is not done by the end of this summer, I suggest the project officially discontinue Zenmap altogether.

**Comment by necrose99** [comment]
personally i'd like to see **zenmap-py3** or a **qt5/[6](https://www.qt.io/blog/2019/08/07/technical-vision-qt-6)**-zenmap port ** as a further enhancement with ability to run nse scripts etc... or PY-QT ui port. 
some of us dyslexic  sobs can use console for simple shit , but sometime the gui is a good bit of help. 

https://www.bountysource.com/issues/56841256-zenmap-and-ndiff-are-python2-only  15 on the issue 
 
thanks to airdrops on key base i have a bit of lose lumen's  needing a good home.

**Comment by necrose99** [comment]
```
#!/usr/bin/env python

# Ndiff
#
# This programs reads two Nmap XML files and displays a list of their
# differences.
#
# Copyright 2008 Insecure.Com LLC
# Ndiff is distributed under the same license as Nmap. See the file COPYING or
# https://nmap.org/data/COPYING. See https://nmap.org/book/man-legal.html for more
# details.
#
# David Fifield
# based on a design by Michael Pattrick

from __future__ import print_function
import sys

# Check if the given directory, and all its parent directories, are owned and
# writable only by our euid or by root. If symlinks are present, they are
# recursively checked, up to a limit of SYMLINK_LIMIT.
# https://www.securecoding.cert.org/confluence/display/seccode/FIO15-C.+Ensure+that+file+operations+are+performed+in+a+secure+directory
# We use this code for Zenmap too
SYMLINK_LIMIT = 5
def is_secure_dir(path, num_symlinks=0):
    import os
    import os.path
    import stat

    if not os.path.isabs(path):
        return False

    if num_symlinks >= SYMLINK_LIMIT:
        return False

    dirs = []
    while True:
        dirs.append(path)
        dirname = os.path.dirname(path)
        if dirname == path:
            break
        path = dirname
    # Traverse root-to-leaf.
    dirs.reverse()

    for dir in dirs:
        if os.path.islink(dir):
            link = os.readlink(dir)
            if not is_secure_dir(link, num_symlinks + 1):
                return False
            continue
        if not os.path.isdir(dir):
            return False
        buf = os.stat(dir)
        if buf.st_uid != os.geteuid() and buf.st_uid != 0:
            return False
        if buf.st_mode & (stat.S_IWGRP | stat.S_IWOTH) != 0:
            return False

    return True

# Add the install_lib directory to sys.path, the list of directories searched
# for modules, but don't do it if the directory or its parents may be writable
# by other users. The following line is replaced by the installation program.
INSTALL_LIB = None
if INSTALL_LIB is not None and is_secure_dir(INSTALL_LIB):
    sys.path.append(INSTALL_LIB) # lgtm[py/unreachable-statement]

try:
    import ndiff
except ImportError as e:
    print("""\
Could not import the ndiff module: %s.
I checked in these directories:""" % repr(e.message), file=sys.stderr)
    for dir in sys.path:
        print("    %s" % dir, file=sys.stderr)
    print("""\
If you installed Ndiff in another directory, you may have to add the
modules directory to the PYTHONPATH environment variable.""", file=sys.stderr)
    sys.exit(1)

import ndiff

if __name__ == "__main__":
    sys.excepthook = ndiff.excepthook
    sys.exit(ndiff.main())
```

**Comment by kulikjak** [comment]
And here is Zenmap migrated to Python 3 and pygobject: #2088. This, together with #1807, #1972 (if we want to keep Python 2 compatibility) and some setup.py and configure changes, should be enough to completely port Nmap to Python 3.

**Comment by anthraxx** [comment]
@bonsaiviking @nnposter any progress on integration of the contributions? I now gonna drop the python2 parts from our distro pacakge now, which is sad :cat2:

**Comment by LucianoHanna** [comment]
There's something that I can help in this issue?

**Comment by chmelarp** [comment]
I did some intensive tests on Ubuntu 20.04 and .10 both as root and ordinary user. It behaves like it did before :)
**You can merge it into master**

**Comment by chmelarp** [comment]
Eventually you can decide to make a separate repo for Zenmap ...?

**Comment by Augustin-FL** [comment]
any update on this subject? It's not possible to use Zenmap on Debian bullseye because of its python2 dependency, we have to use alternatives like nmapsi4 instead.

**Comment by N0rbert** [comment]
@Augustin-FL  

For Debian Bullseye you can get packages from Debian Buster and install them with

```
mkdir -p ~/Downloads/zenmap
cd ~/Downloads/zenmap

wget -c http://deb.debian.org/debian/pool/main/libf/libffi/libffi6_3.2.1-9_amd64.deb
wget -c http://deb.debian.org/debian/pool/main/p/pygobject-2/python-gobject-2_2.28.6-13+b1_amd64.deb
wget -c http://deb.debian.org/debian/pool/main/p/pycairo/python-cairo_1.16.2-1+b1_amd64.deb
wget -c http://deb.debian.org/debian/pool/main/p/python-numpy/python-numpy_1.16.2-1_amd64.deb
wget -c http://deb.debian.org/debian/pool/main/p/pygtk/python-gtk2_2.24.0-5.1+b1_amd64.deb
wget -c http://deb.debian.org/debian/pool/main/n/nmap/zenmap_7.70+dfsg1-6+deb10u1_all.deb

sudo apt-get install ./*.deb
```

**Comment by kulikjak** [comment]
It would be really nice to get any response from some Nmap maintainer. Is there something I/we can do to make this happen?

I understand that time is limited, and these are considerable changes, but even knowing that you e.g. "currently don't have time to look into them" would be better than no response at all.

**Comment by fyodor** [comment]
This is is important, but we have a lot of other important as well.  And some of those issues are harder to workaround than this (where you just need to install Python2). Of the 609 currently open Nmap and Npcap issues, this is currently #11 in priority.

**Comment by themadpsychologist** [comment]
I'm attempting to work on the issue, having found it on BountySource, but expected a higher ratio of syntax revisions and module updates. Frankly, I think there's more work here than $15 covers. Is anyone already working on the problem and, if so, could I perhaps be added to the communication for the project?

**Comment by necrose99** [comment]
@themadpsychologist

sure its not much was hoping others would have added to it ie the bounty  

https://github.com/nmap/nmap/pull/2088

**Comment by themadpsychologist** [comment]
Okay, I [updated Ndiff](https://github.com/nmap/nmap/pull/2274), but the [build tests failed](https://travis-ci.org/github/nmap/nmap/builds/764022714). Help is appreciated.

**Comment by BryanQuigley** [comment]
@themadpsychologist  - there are already working branches for ndiff and one for nmap IIRC, it's stalled on those with commit access not taking them.

**Comment by themadpsychologist** [comment]
@BryanQuigley Thank you very much for your help.

I'm not familiar with working branches; I'm only seeing the `master` branch. Where should I be looking?

By "those with commit access", are you referring to people who have write access to the repository?

Most importantly, what should I do? Should I reopen the pull request, delete it, or fiddle with it somehow?

**Comment by BryanQuigley** [comment]
I don't believe more coding work is needed - it's more organizational - see the open PRs here https://github.com/nmap/nmap/pulls?q=is%3Apr+is%3Aopen+python3

**Comment by themadpsychologist** [comment]
...Oh. Well, then.

Could somebody please remove the BountySource listing, perchance? :(

I should have read the thread before beginning my coding. I learned some things, anyway.

**Comment by N0rbert** [comment]
Possible workaround is to use [Flatpak version of ZenMap](https://flathub.org/apps/details/org.nmap.Zenmap) as follows:

```
sudo apt-get update
sudo apt-get install flatpak

flatpak remote-add --if-not-exists flathub https://flathub.org/repo/flathub.flatpakrepo
flatpak install flathub org.nmap.Zenmap
```

Such installation needs ~330 Mb of downloads, and will use ~1 Gb on disk.

**Comment by sql-sith** [comment]
> Possible workaround is to use [Flatpak version of ZenMap](https://flathub.org/apps/details/org.nmap.Zenmap) as follows:
This worked for me but there is a behavior I don't understand. When I try to use fingerprinting, Zenmap says I need root access to do so. However, when I launch Flatpak/Zenmap as root, I get errors saying `dnet: Failed to open device enx808abdbef5eb`. The device name is correct for the ethernet device is correct, and the nmap command that causes Zenmap (as root) to display this error runs successfully in Terminal (also as root).

I know this isn't the best place to raise this question, but I can't find any definitive help about why this might be happening. Can someone get me pointed to a good resource? Or is there a way to get (verbose) logging from Zenmap? I don't see any logs being touched anywhere in `/var/log`, and I don't see any files in my home directory matching `*zen*` that look like log files either.

**Comment by cclauss** [comment]
https://pypi.org/project/pyndiff

**Comment by medmedin2014** [comment]
For now all Arch based distros removed python2 from their official repos, and it's nightmare to build it from AUR.

**Comment by N0rbert** [comment]
https://github.com/nmap/nmap/issues/1176#issuecomment-1013827740

**Comment by maravento** [comment]
> Use Flatpak then as described in [#1176 (comment)](https://github.com/nmap/nmap/issues/1176#issuecomment-1013827740) .

very heavy

**Comment by chmelarp** [comment]
Please, use this Python3 version
https://github.com/nmap/nmap/pull/2088#issuecomment-1302579968

**Comment by chmelarp** [comment]
Please, use this Python3 version
https://github.com/nmap/nmap/pull/2088#issuecomment-1302579968

**Comment by chmelarp** [comment]
Please, use Python3 version as described here
https://github.com/nmap/nmap/pull/2088#issue-666044617

**Comment by maravento** [comment]
The zenmap flatpack project seems abandoned. The expected results are not obtained.

**Comment by chmelarp** [comment]
Sure, use this branch:
https://github.com/kulikjak/nmap/tree/master-python3

**Comment by cclauss** [comment]
@chmelarp Based on the incomplete kulikjak/nmap#1, the `master-python3` branch needs a lot of work.

I would be interested to work on https://github.com/nmap/nmap/pulls/cclauss if any maintainers of the nmap/nmap repo are willing to review PRs.  It has been 1,051 days since Python 2 end of life on 1/1/2020 so it is past the time to make this upgrade.

**Comment by chmelarp** [comment]
Thank you for your effort, @cclauss !

@kulikjak 's code works OK, but it may need some polishing as he made it quite a time ago. There are plenty of deprecated warnings and the README should be updated from what I can see.

Also, Zenmap is mostly independent on the C code. IMHO it should have a separate repository so it can be packaged and maintained separately.

What is your opinion @kulikjak, @fyodor, @cldrn, @hsluoyz ?

**Comment by kulikjak** [comment]
Hi, I rebased my change pretty recently (3 months ago) and based on my limited testing, everything seemed to work - but I won't deny there might be some polishing to be done. That said, it appears that some people (maybe even distros) are already using it, and I fixed most issues they reported so far.

I didn't get rid of all deprecation warnings because that would mean even more changes, and I didn't want to rewrite it completely as that would make the review and subsequent merge even harder. The same goes for the GTK 4 migration ([I asked about it in January](https://github.com/nmap/nmap/pull/2088#issuecomment-1013837615), but got no upstream response for that and thus didn't really explore it further yet).

`README` and e.g. `setup.py` might need some changes as well, but there were other PRs that looked into that (I believe), and I focused on the Zenmap only. I can certainly look at `README` and `setup.py` update as well.

I guess that separate repository might make sense, considering that the Zenmap GUI isn't really the core of Nmap and doesn't really depend on it that much.

**Comment by darkbasic** [comment]
I'm all in for the separate repo: it makes sense from a coding point of view and it will allow to proceed faster as well.

**Comment by kulikjak** [comment]
> @chmelarp Based on the incomplete [kulikjak#1](https://github.com/kulikjak/nmap/pull/1), the `master-python3` branch needs a lot of work.

Well, I think it's complete from the Zenmap point of view (I mean, except for possible bugs I didn't catch). I didn't touch `ndiff.py` because there is #1807 for that, and `setup.py` or `Makefile.in` because those are not really about Zenmap specifically - but they will certainly need an update as well.

**Comment by cclauss** [comment]
@cldrn @fyodor @hsluoyz Would it be possible to create repos for:
* https://github.com/nmap/ndiff
* https://github.com/nmap/zenmap
so that we can focus on modernizing those codebases?  There are currently almost 500 open issues and more than 200 open pull requests on https://github.com/nmap/nmap so perhaps all would benefit from a bit of divide and conquer.

If there are more appropriate repos to focus on these two codebases, please let us know.
